### DISCUSSION TITLE
How to close a client connection?

### DISCUSSION THREAD

**Discussion by XMethues** [hyperium/tonic#1024] [Q&A] [answered]
I create a "shared client"(not channel) in the app(tauri) that connects to go backend server. An action(like create_ client method) create the shared client and later another action(start_receive_stream) call a server side stream. The question is How can I close the server stream in tonic? In go, I think I can call close to close the connection after dialing, but I can't find the equivalent method in tonic client.
Ideas how to do it.
1. create a channel that break the meessage loop. Select both channel receive and stream message. Once receive close break the for look select.
2. a grpc call, for example, making a grpc call to the server. Server send the canncel status to the client. Client check the message data then break the stream message loop.

By the way, is it ok to just create one client? or should i create a client every time?

**Comment by XMethues** [comment, marked as answer]
I alreay fixed it

**Reply by LucioFranco** [reply]
Could you post how you fixed it for future people?

**Reply by XMethues** [reply]
Using method 2. Basically, I use the same way as how to handle websocket connection on server.
active shutdown
```
func (ts *Service) Close(ctx context.Context, in *CLoseRequest) (*empty.Empty, error) {
       // get your client here based on your code
	client.Done <- true 
	return &emptypb.Empty{}, nil
}
```
go code. In the stream function.
```
	ctx := stream.Context() // very very very important. This will close the stream when the client  crush or any accidents
	for {
		select {
		case <-ctx.Done(): // positive shutdown
			log.Println("client connection disconnected")
			return nil
		case <-client.Done: // active shutdown
			log.Println("client connect ready to close")
			return nil
		case msg := <-client.Message:
			log.Println("receive message", msg)
			if err := stream.Send(msg); err != nil {
				return err
			}
		}
	}
```
